File: src/juliacode/src/topic_classification.jl
using Flux
using CSV
using DataFrames
using TextAnalysis
using Statistics
using Random

# Set a random seed for reproducibility
Random.seed!(1234)

# Load dataset from CSV file
function load_dataset(file_path)
    df = CSV.read(file_path, DataFrame)
    texts = df.text
    labels = [split(label, ", ") for label in df.label]
    return texts, labels
end

# Preprocess text (tokenization, lowercase)
function preprocess_text(text::String)
    tokens = split(lowercase(text), r"\s+")
    return Document(tokens)
end

# Create a vocabulary from documents
function create_vocab(documents)
    return unique(vcat([d.tokens for d in documents]...))
end

# Convert text to TF-IDF vector manually
function text_to_tfidf(documents, vocab)
    tfidf_vectors = []
    for doc in documents
        tfidf_vector = zeros(Float32, length(vocab))
        for (i, word) in enumerate(vocab)
            tfidf_vector[i] = tf(doc, word) * idf(documents, word)
        end
        push!(tfidf_vectors, tfidf_vector)
    end
    return hcat(tfidf_vectors...)
end

# TF-IDF calculation
function tf(doc, word)
    count = sum(w == word for w in doc.tokens)
    return count / length(doc.tokens)
end

function idf(documents, word)
    doc_count = sum(word in d.tokens for d in documents)
    return log(length(documents) / (1 + doc_count))
end

# Define the Model
function create_model(input_size, num_topics)
    model = Chain(
        Dense(input_size, 128, relu),  # Input layer with input size equal to TF-IDF vector length
        Dense(128, 64, relu),          # Hidden layer with 64 neurons
        Dense(64, num_topics),         # Output layer with number of topics
        sigmoid                        # Sigmoid activation for multi-label classification
    )
    return model
end

# Train the Model
function train_model(documents, labels, vocab, topic_labels)
    model = create_model(length(vocab), length(topic_labels))

    # Convert documents to TF-IDF vectors
    x_train = text_to_tfidf(documents, vocab)

    # Convert labels to one-hot encoding using actual topic labels
    y_train = zeros(Float32, length(topic_labels), length(labels))
    for (i, label_set) in enumerate(labels)
        for label in label_set
            idx = findfirst(isequal(label), topic_labels)
            if idx !== nothing
                y_train[idx, i] = 1.0
            end
        end
    end

    # Define loss function and optimizer
    loss(x, y) = Flux.logitcrossentropy(model(x), y)
    opt = ADAM()

    # Training loop
    for epoch in 1:50  # Adjust epochs as needed
        Flux.train!(loss, Flux.params(model), [(x_train, y_train)], opt)
        println("Epoch $epoch: Loss = $(loss(x_train, y_train))")
    end

    return model
end

# Predict Topics
function predict_topics(text::String, model, vocab, topic_labels)
    processed_text = preprocess_text(text)
    tfidf_vector = text_to_tfidf([processed_text], vocab)[:, 1]
    pred = model(tfidf_vector)
    
    # Print predictions for debugging
    println("Predicted probabilities: ", pred)

    # Adjust threshold if needed
    threshold = 0.5
    pred_topics = topic_labels[pred .> threshold]
    
    # Print topics that exceed the threshold
    println("Predicted topics based on threshold: ", pred_topics)
    
    return pred_topics
end

function main()
    # Load dataset
    texts, labels = load_dataset("topic_classification.csv")
    
    # Preprocess texts
    processed_texts = [preprocess_text(text) for text in texts]

    # Create a vocabulary from all texts
    vocab = create_vocab(processed_texts)

    # Get unique topic labels from the dataset
    topic_labels = unique(vcat(labels...))

    # Train the model
    model = train_model(processed_texts, labels, vocab, topic_labels)

    # Example input
    text = "Hello guys this is my post on mazaryn social network after reading some books and playin Guitar"
    topics = predict_topics(text, model, vocab, topic_labels)
    println("Predicted Topics: ", topics)
end

main()

File: src/juliacode/src/topic_classification.py
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.model_selection import train_test_split
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Load dataset from CSV file
def load_dataset(file_path):
    df = pd.read_csv(file_path)
    texts = df['text']
    labels = df['label'].apply(lambda x: [label.strip() for label in x.split(",")])
    return texts, labels

# Define the Model
def create_model(input_size, num_topics):
    model = Sequential([
        Dense(128, activation='relu', input_shape=(input_size,)),
        Dense(64, activation='relu'),
        Dense(num_topics, activation='sigmoid')
    ])
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

# Train the Model
def train_model(X_train, y_train, input_size, num_topics):
    model = create_model(input_size, num_topics)
    model.fit(X_train, y_train, epochs=50, verbose=1)
    return model

# Predict Topics
def predict_topics(text, model, vectorizer, mlb):
    tfidf_vector = vectorizer.transform([text])
    pred = model.predict(tfidf_vector)
    
    # Print raw prediction probabilities for debugging
    print("Raw prediction probabilities: ", pred)
    
    # Adjust threshold if needed
    threshold = 0.1  # Lower threshold for testing
    pred_topics = mlb.classes_[np.where(pred > threshold)[1]]
    
    return pred_topics

def main():
    # Load dataset
    texts, labels = load_dataset("topic_classification.csv")

    # Create a TF-IDF vectorizer and transform the texts
    vectorizer = TfidfVectorizer()
    X = vectorizer.fit_transform(texts)
    
    # Convert labels to binary format
    mlb = MultiLabelBinarizer()
    y = mlb.fit_transform(labels)

    # Split data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1234)

    # Train the model
    num_topics = len(mlb.classes_)
    model = train_model(X_train, y_train, X.shape[1], num_topics)

    # Example input
    text = "Hello guys this is my post on mazaryn social network after reading some books and playing Guitar"
    topics = predict_topics(text, model, vectorizer, mlb)
    print("Predicted Topics:", topics)

if __name__ == "__main__":
    main()
